File: epi_judge_python/run_length_compression.py
from test_framework import generic_test
from test_framework.test_failure import TestFailure


def decoding(s: str) -> str:
    count, res = 0, []
    for ch in s:
        if ch.isdigit():
            count = 10 * count + int(ch)
        else:
            res.append(count * ch)
            count = 0
    return ''.join(res)


def encoding(s: str) -> str:
    res = ''
    count = 1
    if len(s) < 1:
        return ''
    prev = s[0]
    for i in range(1, len(s)):
        if s[i] == prev:
            count += 1
        else:
            res += str(count) + s[i]
            prev = s[i]
            count = 0
    return res


def rle_tester(encoded, decoded):
    if decoding(encoded) != decoded:
        raise TestFailure('Decoding failed')
    if encoding(decoded) != encoded:
        raise TestFailure('Encoding failed')


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('run_length_compression.py',
                                       'run_length_compression.tsv',
                                       rle_tester))

File: epi_judge_cpp/run_length_compression.cc
#include <cctype>
#include <string>

#include "test_framework/generic_test.h"
#include "test_framework/test_failure.h"
using std::string;
string Decoding(const string &s) {
  int count{0};
  string result;
  for (auto i = 0; i < s.size(); i++) {
    if (std::isdigit(s[i])) {
      count = count * 10 + (s[i] - '0');
    } else {
      for (auto j = 0; j < count; j++) {
        result.push_back(s[i]);
      }
      count = 0;
    }
  }
  return result;
}
string Encoding(const string &s) {
  string result;
  char seen = s.front();
  int count = 0;
  for (auto &ch : s) {
    if (ch == seen) {
      count++;
    } else {
      result.append(std::to_string(count));
      result.push_back(seen);
      seen = ch;
      count = 1;
    }
  }
  result.append(std::to_string(count));
  result.push_back(seen);
  return result;
}
void RleTester(const string &encoded, const string &decoded) {
  if (Decoding(encoded) != decoded) {
    throw TestFailure("Decoding failed");
  }
  if (Encoding(decoded) != encoded) {
    throw TestFailure("Encoding failed");
  }
}

int main(int argc, char *argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"encoded", "decoded"};
  return GenericTestMain(args, "run_length_compression.cc",
                         "run_length_compression.tsv", &RleTester,
                         DefaultComparator{}, param_names);
}
